### ISSUE TITLE
Trouble installing custom version of DyNet with Boost

### ISSUE THREAD

**Issue by jbrry** [deep-spin/lp-sparsemap#3]
Hi Vlad, thank you for the exciting resource. Looking forward to trying it out!

I am trying to install all of the dependencies. I am having trouble getting DyNet installed using the [DyNet fork](https://github.com/vene/dynet).

I have followed the instructions to install DyNet for C++ here: https://dynet.readthedocs.io/en/latest/install.html#building

Where I have diverged from the official instructions is instead of cloning the official DyNet repo: 
`git clone https://github.com/clab/dynet.git`

I am cloning your custom version:
`git clone https://github.com/vene/dynet.git`.

I am then switching to the `py3` branch: `git checkout py3`. 

I am then running the cmake command as normal: 

```
(base) jbarry@jbarry-desktop:~/dynet-base3/dynet/build$ cmake .. -DEIGEN3_INCLUDE_DIR=/home/jbarry/dynet-base3/eigen -DENABLE_CPP_EXAMPLES=ON
-- Boost dir is Boost_INCLUDE_DIR-NOTFOUND
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find Boost (missing: Boost_INCLUDE_DIR program_options regex
  serialization)
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:2179 (find_package_handle_standard_args)
  CMakeLists.txt:103 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/jbarry/dynet-base3/dynet/build/CMakeFiles/CMakeOutput.log".
``` 

I then tried downloading [Boost v 1.62](https://sourceforge.net/projects/boost/files/boost/1.62.0/) which seemed to be the most recent version of Boost at the time of the Dynet fork around November 2016 and tried again using the [instructions for compiling with Boost](https://dynet.readthedocs.io/en/latest/install.html#compiling-with-boost):

```
(lp-sparsemap-pt-dy-custom) jbarry@jbarry-desktop:~/dynet-base3/dynet/build$ cmake .. -DEIGEN3_INCLUDE_DIR=/home/jbarry/dynet-base3/eigen -DENABLE_CPP_EXAMPLES=ON -DENABLE_BOOST=ON
-- Boost dir is /home/jbarry/dynet-base3/boost
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find Boost (missing: program_options regex serialization) (found
  version "1.62.0")
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:2179 (find_package_handle_standard_args)
  CMakeLists.txt:103 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/jbarry/dynet-base3/dynet/build/CMakeFiles/CMakeOutput.log".
```

I am probably doing something completely wrong when compiling with Boost. Apologies as I am quite inexperienced with using `cmake` commands and editing CMake options. Here is how my DyNet base directory looks:

```
(lp-sparsemap-pt-dy-custom) jbarry@jbarry-desktop:~/dynet-base3$ ls
boost dynet  eigen

ls dynet/
build  cmake  CMakeLists.txt  config.h.cmake  doc  dynet  examples  external  LICENSE  MEM.notes  pyexamples  python  python_tests  README.md  rnnlm  tests  tutorial
```

I'm just wondering if you have any pointers to install this correctly? I was able to successfully install the official DyNet code in a separate environment without Boost. Attached is the `CMakeFiles/CMakeOutput.log` if it helps. Many thanks!
[CMakeOutput.log](https://github.com/deep-spin/lp-sparsemap/files/4947978/CMakeOutput.log)

**Comment by jbrry** [comment]
I'm sorry, I misinterpreted the meaning of "standard location" in the DyNet docs (I thought the boost folder had to be placed in the base directory along with `eigen`). The problem was fixed with installing boost on my system:

`sudo apt-get install libboost-all-dev`

**Comment by vene** [comment]
Happy this worked-- I want to add that it should be possible to build dynet without boost at all (this is how I've been doing it). I think there is a cmake flag for thism

**Comment by jbrry** [comment]
Hey Vlad, thanks! And I also made the mistake of building on the `py3` branch instead of `targets` (that might have something to do with why boost was needed). After I installed Dynet with boost, I realised I had used the wrong branch :) so I ran everything again and built the tool with no errors. I will try to run a few of the examples and see how I get on and will open a separate issue (to keep the issues on topic) if I get stuck!
